contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -81,860 +81,1019 @@ namespace net_utils
+                on_interrupted();
+              else if (state.status == status_t::TERMINATING)
+                on_terminating();
+            }
+            else if (ec.value()) {
+              terminate();
+            }
+            else if (
+              not epee::net_utils::is_ssl(
+                static_cast<const unsigned char *>(
+                  state.data.read.buffer.data()
+                ),
+                bytes_transferred
+              )
+            ) {
+              state.ssl.enabled = false;
+              start_read();
+            }
+            else {
+              state.ssl.detected = true;
+              start_handshake();
+            }
+          }
+        )
+      );
+      return;
+    }
 
-    boost::system::error_code ec;
-    auto remote_ep = socket().remote_endpoint(ec);
-    CHECK_AND_NO_ASSERT_MES(!ec, false, "Failed to get remote endpoint: " << ec.message() << ':' << ec.value());
-    CHECK_AND_NO_ASSERT_MES(remote_ep.address().is_v4() || remote_ep.address().is_v6(), false, "only IPv4 and IPv6 supported here");
+    state.socket.wait_handshake = true;
+    auto on_handshake = [this, self](const ec_t &ec, size_t bytes_transferred){
+      lock_guard_t guard(state.lock);
+      state.socket.wait_handshake = false;
+      if (state.socket.cancel_handshake) {
+        state.socket.cancel_handshake = false;
+        if (state.status == status_t::RUNNING)
+          interrupt();
+        else if (state.status == status_t::INTERRUPTED)
+          on_interrupted();
+        else if (state.status == status_t::TERMINATING)
+          on_terminating();
+      }
+      else if (ec.value()) {
+        ec_t ec;
+        connection_basic::socket_.next_layer().shutdown(
+          socket_t::shutdown_both,
+          ec
+        );
+        connection_basic::socket_.next_layer().close(ec);
+        state.socket.connected = false;
+        interrupt();
+      }
+      else {
+        state.ssl.handshaked = true;
+        start_write();
+        start_read();
+      }
+    };
+    const auto handshake = handshake_t::server;
+    static_cast<shared_state&>(
+      connection_basic::get_state()
+    ).ssl_options().configure(connection_basic::socket_, handshake);
+    strand.post(
+      [this, self, on_handshake]{
+        connection_basic::socket_.async_handshake(
+          handshake,
+          boost::asio::buffer(
+            state.data.read.buffer.data(),
+            epee::net_utils::get_ssl_magic_size()
+          ),
+          strand.wrap(on_handshake)
+        );
+      }
+    );
+  }
 
-    if (remote_ep.address().is_v4())
-    {
-      const unsigned long ip_ = boost::asio::detail::socket_ops::host_to_network_long(remote_ep.address().to_v4().to_ulong());
-      return start(is_income, is_multithreaded, ipv4_network_address{uint32_t(ip_), remote_ep.port()});
+  template<typename T>
+  void connection<T>::start_read()
+  {
+    if (state.timers.throttle.in.wait_expire || state.socket.wait_read ||
```

Comment:
`wait_read` is only necessary because the callbacks invoke `start_read` ... again. This is unnecessary if the callback were a local struct that contains only the inner looping read calls. I think this is preferable because it separates the start and looping code, whereas this is mixing the two. It also allows for less state machines.

Reply:
It will be needed later when everything is asynchronous since reader may not consume every received chunk immediately and it's possible that there is no active callbacks at all.

Reply:
The flag is only needed because `start_read` is both starting and continuing the reads. It is possible to separate the two, removing the need for the flag. After every async read (and subsequent processing), the continue function is called.

Reply:
I didn't try to optimize number of states / transitions.
It can be done later unless it's needed for significant performance increase or it's related to correctness of the code.

Reply:
This is just my personal opinion obviously, but -

if you want to do a re-write of the entire connection code, then it should be "optimized" with the fewest number of states / transitions possible. Otherwise, what is the purpose of the re-write? I'm being critical because if this were done piecemeal, the code could've looked similar in the end to this PR, but the unnecessary code/states could be removed along the way easier during review. Removing the unnecessary bits makes it easier to maintain and debug.

As an example, the `cancel_` state flags have to be removed, unless there is some justification why the standard way of checking for cancellation is inappropriate (none was provided above in the other comment thread). And this `wait_read`  (that started this thread) doesn't appear to have any justification (yet), given how ASIO callbacks are typically done.

Reply:
> Otherwise, what is the purpose of the re-write?
>
These explicit flags are needed to not depend on internal behaviour of boost asio which doesn't allow to cancel already dispatched callback.

Reply:
> ... Removing the unnecessary bits makes it easier to maintain and debug.
>
It depends.

Reply:
>if you want to do a re-write of the entire connection code, then it should be "optimized" with the fewest number of states / transitions possible. Otherwise, what is the purpose of the re-write? 
>
The main goal is to fix known problems and prepare code for design changes, which can't be done in case of greedy early optimizations of every component.

Reply:
> These explicit flags are needed to not depend on internal behaviour of boost asio which doesn't allow to cancel already dispatched callback.

I've thought about this case before, and typically its not worth stressing about. In the read/write callbacks either reset the timer immediately or call `cancel` immediately. The timer will be in the strand queue with the appropriate `error_code`.

In the off chance that it queued the timer just after the other event but before the cancel could occur  ... the timer was so close to executing its not worth the extra state variable and logic. Let ASIO handle all of the state tracking because its better "defensive coding" - the `error_code` cancel check handles every future update to how the timer is used.

> It depends.

Just chop out these unnecessary states. There's a giant `if` ladder than get be chopped down for instance.

> The main goal is to fix known problems and prepare code for design changes, which can't be done in case of greedy early optimizations of every component.

I dislike your usage of the word optimization in this context because it makes it sound like I am asking for a performance comparison. This is about making the code more concise+legible - particularly the `start_read` function.

Reply:
"concise - short and clear, expressing what needs to be said without unnecessary words"
"optimization - the process of making something as good or effective as possible"
Non-programming language is ambiguous enough to allow misinterpretation of anything so If current code is correct then any further changes can be treated as optimization.

Reply:
You're trying to avoid making changes, which is frustrating because this PR contains more code than necessary, leaving more work for reviewers. I will try again in a new review to point out some things.